Assets/LeapMotionModules/Interaction/InteractionManager.cs

```diff
@@ -0,0 +1,650 @@
+            });
+        }
+
+        // Check sustained grasping.
+        if (enableContact) {
+          RemapInteractionObjectStateChecks(
+            hands: _interactionHands,
+            stateCheckFunc: (InteractionHand maybeSustainedGraspingHand, out IInteractionBehaviour graspedObject) => {
+              return maybeSustainedGraspingHand.CheckGraspHold(out graspedObject);
+            },
+            actionPerInteractionObject: (contactedObject, contactingIntHands) => {
+              contactedObject.StayGrasped(contactingIntHands);
+            });
+        }
+
+      }
+    }
+    
+    private delegate V StateChangeCheckFunc<H, T, V>(H hand, out T obj);
+    private delegate V MultiStateChangeCheckFunc<H, T, V>(H hand, out HashSet<T> objs);
+    
+    private static Pool<List<InteractionHand>> s_intHandsListPool = new Pool<List<InteractionHand>>();
+    private static Dictionary<IInteractionBehaviour, List<InteractionHand>> s_objHandsMap = new Dictionary<IInteractionBehaviour, List<InteractionHand>>();
+
+    /// <summary>
+    /// Checks object state per-hand, then calls an action per-object with all hand checks that reported back an object.
+    /// </summary>
+    private static void RemapInteractionObjectStateChecks(
+        InteractionHand[]                                                   hands,
+        StateChangeCheckFunc<InteractionHand, IInteractionBehaviour, bool>  stateCheckFunc,
+        Action<IInteractionBehaviour, List<InteractionHand>>                actionPerInteractionObject) {
+
+      // Ensure the object->hands buffer is clean.
+      s_objHandsMap.Clear();
+
+      // In a nutshell, this remaps methods per-hand that output an interaction object if the hand changed that object's state
+      // to methods per-object with all of the hands for which the check produced a state-change.
+      foreach (var hand in hands) {
+        IInteractionBehaviour objectWhoseStateChanged;
+        if (stateCheckFunc(hand, out objectWhoseStateChanged)) {
+          if (!s_objHandsMap.ContainsKey(objectWhoseStateChanged)) {
+            s_objHandsMap[objectWhoseStateChanged] = s_intHandsListPool.Take();
+          }
+          s_objHandsMap[objectWhoseStateChanged].Add(hand);
+        }
+      }
+      // Finally, iterate through each (object, hands) pair and call the action for each pair
+      foreach (var objHandsPair in s_objHandsMap) {
+        actionPerInteractionObject(objHandsPair.Key, objHandsPair.Value);
+
+        // Clear each hands list and return it to the list pool.
+        objHandsPair.Value.Clear();
+        s_intHandsListPool.Return(objHandsPair.Value);
+      }
+    }
+
+    /// <summary>
+    /// Checks object state per-hand, then calls an action per-object with all hand checks that reported back objects.
+    /// </summary>
+    private static void RemapMultiInteractionObjectStateChecks(
+        InteractionHand[]                                                        hands,
```

Comment:
ehhh just the sake of purity? It's a static transformation of methods. If it's bothersome I can change it